[iOS] Programmatic focus triggered while evaluating user script should present the software keyboard

whsieh opened the pull request:
#### 2b0c968051a7552d2f54d39bf790e0970fce259d
<pre>
[iOS] Programmatic focus triggered while evaluating user script should present the software keyboard
<a href="https://bugs.webkit.org/show_bug.cgi?id=243416">https://bugs.webkit.org/show_bug.cgi?id=243416</a>
rdar://97945400

Reviewed by Aditya Keerthi and Devin Rousso.

Allow `-evaluateJavaScript:completionHandler:` and `-callAsyncJavaScript:…completionHandler:` to
set the `m_userIsInteracting` flag on WebPage during the scope of user script evaluation. This
allows apps (linked on or after certain versions of iOS) to use `element.focus()` to properly focus
text fields and other form controls, and show the software keyboard in the process.

Test: KeyboardInputTests.InputSessionWhenEvaluatingJavaScript

* Source/WTF/wtf/cocoa/RuntimeApplicationChecksCocoa.h:

Add a new SDK-aligned behavior flag to guard this behavior behind a linked-on-or-after check.

* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::runJavaScript):

When executing user script (i.e. from the WebKit client), set the `m_userIsInteracting` flag to true
as long as the app is linked on or after the SDK versions corresponding to
`SDKAlignedBehavior::ProgrammaticFocusDuringUserScriptShowsInputViews`.

* Tools/TestWebKitAPI/Tests/ios/KeyboardInputTestsIOS.mm:

Add an API test to verify that using `-_evaluateJavaScriptWithoutUserGesture:…:` to evaluate script
that focuses a text field does not cause an input session to start; however, using either
`-callAsyncJavaScript:…:` or `-evaluateJavaScript:…:` starts an input session.

Canonical link: <a href="https://commits.webkit.org/253034@main">https://commits.webkit.org/253034@main</a>
</pre>

dcrousso commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -3861,6 +3861,15 @@ void WebPage::runJavaScript(WebFrame* frame, RunJavaScriptParameters&& parameter
+        return m_userIsInteracting;
+    }();
+
+    SetForScope userIsInteractingChange { m_userIsInteracting, shouldAllowUserInteraction };
```
do we also want to check `parameters.forceUserGesture`?  i think that `-[WKWebView _evaluateJavaScriptWithoutUserGesture:completionHandler:]` would probably not want to set the user interaction flag 😅

(part of me wants to ask you to move this to `ScriptController::executeScriptInWorld` since that's where the `UserGestureIndicator` is, but i understand that would involve lotsa `PageClient` plumbing so it's probably fine here)

whsieh replied:
> do we also want to check `parameters.forceUserGesture`?

Good catch! Yes, I'll check that flag as well (and add an API test to exercise this scenario).

> move this to `ScriptController::executeScriptInWorld`

Well, `m_userIsInteracting` is already a client-layer concept, so I don't think it would be an improvement to add all the client plumbing needed to make that work.

(I think it's also cleaner to be able to use `SetForScope`, as a variable on the stack, so we don't need to worry about setting and unsetting the bit)

dcrousso commented on Source/WTF/wtf/cocoa/RuntimeApplicationChecksCocoa.cpp:
```diff
@@ -172,6 +172,14 @@ static SDKAlignedBehaviors computeSDKAlignedBehaviors()
     if (linkedBefore(dyld_spring_2022_os_versions, DYLD_IOS_VERSION_15_4, DYLD_MACOSX_VERSION_12_3))
         disableBehavior(SDKAlignedBehavior::AuthorizationHeaderOnSameOriginRedirects);
 
+    if (linkedBefore(dyld_fall_2022_os_versions, DYLD_IOS_VERSION_16_0, DYLD_MACOSX_VERSION_13_0)) {
```
i feel like this might be better as a separate change 😅

whsieh replied:
I can split it out, sure.

whsieh committed:
[iOS] Programmatic focus triggered while evaluating user script should present the software keyboard
https://bugs.webkit.org/show_bug.cgi?id=243416
rdar://97945400

Reviewed by Aditya Keerthi and Devin Rousso.

Allow `-evaluateJavaScript:completionHandler:` and `-callAsyncJavaScript:…completionHandler:` to
set the `m_userIsInteracting` flag on WebPage during the scope of user script evaluation. This
allows apps (linked on or after certain versions of iOS) to use `element.focus()` to properly focus
text fields and other form controls, and show the software keyboard in the process.

Test: KeyboardInputTests.InputSessionWhenEvaluatingJavaScript

* Source/WTF/wtf/cocoa/RuntimeApplicationChecksCocoa.h:

Add a new SDK-aligned behavior flag to guard this behavior behind a linked-on-or-after check.

* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::runJavaScript):

When executing user script (i.e. from the WebKit client), set the `m_userIsInteracting` flag to true
as long as the app is linked on or after the SDK versions corresponding to
`SDKAlignedBehavior::ProgrammaticFocusDuringUserScriptShowsInputViews`.

* Tools/TestWebKitAPI/Tests/ios/KeyboardInputTestsIOS.mm:

Add an API test to verify that using `-_evaluateJavaScriptWithoutUserGesture:…:` to evaluate script
that focuses a text field does not cause an input session to start; however, using either
`-callAsyncJavaScript:…:` or `-evaluateJavaScript:…:` starts an input session.

Canonical link: https://commits.webkit.org/253034@main
